dot/state/grandpa_changes.go

```diff
@@ -0,0 +1,313 @@
+		if err != nil {
+			return false, fmt.Errorf("cannot check ancestry: %w", err)
+		}
+
+		return isDescendant && effectiveNumber == importedNumber, nil
+	})
+
+}
+
+// lookupChangeWhere return the first pending change which satisfy the condition
+func (oc orderedPendingChanges) lookupChangeWhere(condition conditionFunc[*pendingChange]) (
+	pendingChange *pendingChange, err error) {
+	for _, change := range oc {
+		ok, err := condition(change)
+		if err != nil {
+			return nil, fmt.Errorf("failed while applying condition: %w", err)
+		}
+
+		if ok {
+			return change, nil
+		}
+	}
+
+	return nil, nil //nolint:nilnil
+}
+
+// importChange only tracks the pending change if and only if it is the
+// unique forced change in its fork, otherwise will return an error
+func (oc *orderedPendingChanges) importChange(pendingChange *pendingChange, isDescendantOf isDescendantOfFunc) error {
+	announcingHeader := pendingChange.announcingHeader.Hash()
+
+	for _, change := range *oc {
+		changeBlockHash := change.announcingHeader.Hash()
+
+		if changeBlockHash.Equal(announcingHeader) {
+			return fmt.Errorf("%w: %s", errDuplicateHashes, changeBlockHash)
+		}
+
+		isDescendant, err := isDescendantOf(changeBlockHash, announcingHeader)
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			return fmt.Errorf("%w: for block hash %s", errAlreadyHasForcedChange, changeBlockHash)
+		}
+	}
+
+	orderedChanges := *oc
+
+	// Use a binary search to include the pending change in the right position
+	// of a slice ordered by the effective number and by announcing header number
+	idxToInsert := sort.Search(oc.Len(), func(i int) bool {
+		return orderedChanges[i].effectiveNumber() >= pendingChange.effectiveNumber() &&
+			orderedChanges[i].announcingHeader.Number >= pendingChange.announcingHeader.Number
+	})
+
+	orderedChanges = append(orderedChanges, pendingChange)
+	copy(orderedChanges[idxToInsert+1:], orderedChanges[idxToInsert:])
+	orderedChanges[idxToInsert] = pendingChange
+	*oc = orderedChanges
+
+	return nil
+}
+
+// pruneChanges will remove changes whose are not descendant of the hash argument
+// this function updates the current state of the change tree
+func (oc *orderedPendingChanges) pruneChanges(hash common.Hash, isDescendantOf isDescendantOfFunc) error {
+	onBranchForcedChanges := make([]*pendingChange, 0, oc.Len())
+
+	for _, forcedChange := range *oc {
+		isDescendant, err := isDescendantOf(hash, forcedChange.announcingHeader.Hash())
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			onBranchForcedChanges = append(onBranchForcedChanges, forcedChange)
+		}
+	}
+
+	*oc = onBranchForcedChanges
+	return nil
+}
+
+type pendingChangeNode struct {
+	change *pendingChange
+	nodes  []*pendingChangeNode
+}
+
+func (c *pendingChangeNode) importNode(blockHash common.Hash, blockNumber uint, pendingChange *pendingChange,
```

Comment:
this function is a little difficult to read knowing that it calls itself recursively, and with all the different return cases.  Is there a way to refactor this to clearly show all the return cases.  If a `pendingChangeNode` has children and the supplied `pendingChange` not successfully imported from one of the children, then it's supposed to append it as a child? I'd consider refactoring this code or putting some more comments.